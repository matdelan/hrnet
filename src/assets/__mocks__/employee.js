export const mockEmployeeResponse = {
    id:'1',
    firstname: 'John', 
    lastname: 'Doe',    
    dateBirthday: '1990-01-01', 
    dateStart: '2022-09-01',    
    street: '123 Main St', 
    city: 'Springfield',   
    state: 'Illinois',   
    zipCode: '62704',    
    department: 'Marketing',    
}

export const mockEmployeeNew = {
    firstname: 'John', 
    lastname: 'Doe',    
    dateBirthday: '1990-01-01', 
    dateStart: '2022-09-01',    
    street: '123 Main St', 
    city: 'Springfield',   
    state: 'Illinois',   
    zipCode: '62704',    
    department: 'Marketing',    
}