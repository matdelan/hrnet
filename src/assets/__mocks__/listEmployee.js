export const mockListEmployee = {
    status:'200',
    message:'employee created successfully',
    data: [
        {
            id:'1',
            firstname: 'John', 
            lastname: 'Doe',    
            dateBirthday: '1990-01-01', 
            dateStart: '2022-09-01',    
            street: '123 Main St', 
            city: 'Springfield',   
            state: 'Illinois',   
            zipCode: '62704',    
            department: 'Marketing',  
        } ,
        {
            id:'2',
            firstname: 'Johna', 
            lastname: 'aDoe',    
            dateBirthday: '1990-01-01', 
            dateStart: '2022-09-01',    
            street: '123 Main St', 
            city: 'Springfield',   
            state: 'Illinois',   
            zipCode: '62704',    
            department: 'Marketing',  
        },
    ],
}