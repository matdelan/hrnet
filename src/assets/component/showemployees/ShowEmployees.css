.activePage {
    border: solid 3px black
}

.table__header {
    display: flex;
    justify-content: space-between;
}

.table__header-elem {
    display: flex;
}

.table__col {
    display: flex;
}

.table__col-block {
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin-left: 3px;
    margin-right: 4px;
}

.table__col-arrow {
    color: grey;
}

.table__pagination {
    display: flex;
    justify-content: flex-end;
    align-items: center;
    margin-top: 20px;
}

.table__pagination-space {
    margin-left: 35px;
    margin-right: 35px;
    cursor: pointer;
}

.table__footer {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: 30px;
}

.fa-caret-up {
    margin-bottom: -4px;
}
.fa-caret-down {
    margin-top: -4px;
}

.table__row-grey{
    background-color: lightgrey;
}
.table__row-white{
    background-color: white;
}
.table__row-block{
    max-width: 150px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}