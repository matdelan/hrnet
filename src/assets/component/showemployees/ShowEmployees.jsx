import React, { useState, useEffect, useRef } from 'react';
import Input from '../input/Input'
import './ShowEmployees.css'

export default function ShowEmployees({ employees }) {
    const [sortedEmployees, setSortedEmployees] = useState([]);
    const [sortConfig, setSortConfig] = useState({ key: null, direction: 'ascending' });

    const inputSearchRef = useRef(null)
    const errorSearchRef = useRef(null)
    const numElementsRef = useRef(0)

    /* Construct table */
    const [currentLine, setCurrentLine] = useState(0)
    const [currentPage, setCurrentPage] = useState(1)
    const [numPage, setNumPage] = useState(0)

    useEffect(() => {
        setSortedEmployees(employees);
    }, [employees]);

    const updateSearch = () => {
        sortData(sortConfig)
    }
    const updateTable = () => {
        setCurrentPage(1)
        setCurrentLine(0)
        sortData(sortConfig)
    }
    const handlePageChange = (page) => {
        setCurrentPage(page);
        setCurrentLine((page*(numElementsRef.current.value)-numElementsRef.current.value))
        sortData(sortConfig)
    };

    const handlePreviousPage = () => {
        const prevPage = currentPage - 1
        if(prevPage !== 0)
        {
            setCurrentPage(prevPage)
            setCurrentLine((prevPage*(numElementsRef.current.value)-numElementsRef.current.value))
            sortData(sortConfig)
        }
            
    }
    const handleNextPage = () => {
        const nextPage = currentPage + 1
        if(nextPage !== (Math.ceil(sortedEmployees.length / numElementsRef.current.value)+1))
        {
            setCurrentPage(nextPage)
            setCurrentLine((nextPage*(numElementsRef.current.value)-numElementsRef.current.value))
            sortData(sortConfig)
        }    
    }
    const renderPagination = () => {
        const pages = []
        const totalPages = Math.ceil(sortedEmployees.length / numElementsRef.current.value)
        pages.push(<div onClick={handlePreviousPage} className='table__pagination-space'>Previous</div>)
        for (let i = 1; i <= totalPages; i++) {
          pages.push(
            <button
              key={i}
              onClick={() => handlePageChange(i)}
              className={i === currentPage ? 'activePage' : ''}
            >
              {i}
            </button>
          )
        }
        pages.push(<div onClick={handleNextPage} className='table__pagination-space'>Next</div>)
    
        let maxElem = (currentLine+parseInt(numElementsRef.current.value))
        if (maxElem > sortedEmployees.length)
            maxElem = sortedEmployees.length
        return <>
            <div className='table__footer'>
                <div> Showing {currentLine+1} of {maxElem} of {sortedEmployees.length}  </div>
                <div className='table__pagination'>{pages}</div>
            </div>
        </> 
    }

    const sortData = (key, direction) => {
        const searchTerm = inputSearchRef.current.value.toLowerCase();
        const filteredEmployees = employees.filter((item) =>
            Object.values(item).some((data) => {
                return String(data).toLowerCase().includes(searchTerm)
            })
        )
    
        const sortedData = [...filteredEmployees].sort((a, b) => {
            if (a[key] < b[key]) {
                return direction === 'ascending' ? -1 : 1
            }
            if (a[key] > b[key]) {
                return direction === 'ascending' ? 1 : -1
            }
            return 0
        })
    
        setSortedEmployees(sortedData);
        setSortConfig({ key, direction })
    };

    return (
        <>
        <div className='table__header'>
            <div className='table__header-elem'>
                <label htmlFor="elementsPage">Éléments par page : </label>
                <select id="elementsPage" onChange={updateTable} ref={numElementsRef}>
                    <option value="10">10</option>
                    <option value="20">20</option>
                    <option value="50">50</option>
                    <option value="100">100</option>
                </select>
            </div>
            <div className='table__header-elem'>
                <label htmlFor="search">Search :</label>
                <input type="text" id="search" ref={inputSearchRef} onChange={updateSearch}/>
            </div>
        </div>
        <table>
            <thead>
                <tr className='table__col-gen'>
                    <th><div className='table__col'>First Name<span className='table__col-block'><i className="fa-solid fa-caret-up table__col-arrow" onClick={() => sortData('firstname','ascending')}></i><i className="fa-solid fa-caret-down table__col-arrow" onClick={() => sortData('firstname','descending')}></i></span></div></th>
                    <th><div className='table__col'>Last Name<div className='table__col-block'><i className="fa-solid fa-caret-up table__col-arrow" onClick={() => sortData('lastname','ascending')}></i><i className="fa-solid fa-caret-down table__col-arrow" onClick={() => sortData('lastname','descending')}></i></div></div></th>
                    <th><div className='table__col'>Date of Birth<div className='table__col-block'><i className="fa-solid fa-caret-up table__col-arrow" onClick={() => sortData('dateBirthday','ascending')}></i><i className="fa-solid fa-caret-down table__col-arrow" onClick={() => sortData('dateBirthday','descending')}></i></div></div></th>
                    <th><div className='table__col'>Start Date<div className='table__col-block'><i className="fa-solid fa-caret-up table__col-arrow" onClick={() => sortData('dateStart','ascending')}></i><i className="fa-solid fa-caret-down table__col-arrow" onClick={() => sortData('dateStart','descending')}></i></div></div></th>
                    <th><div className='table__col'>Department<div className='table__col-block'><i className="fa-solid fa-caret-up table__col-arrow" onClick={() => sortData('departement','ascending')}></i><i className="fa-solid fa-caret-down table__col-arrow" onClick={() => sortData('departement','descending')}></i></div></div></th>
                    <th><div className='table__col'>Street<div className='table__col-block'><i className="fa-solid fa-caret-up table__col-arrow" onClick={() => sortData('street','ascending')}></i><i className="fa-solid fa-caret-down table__col-arrow" onClick={() => sortData('street','descending')}></i></div></div></th>
                    <th><div className='table__col'>City<div className='table__col-block'> <i className="fa-solid fa-caret-up table__col-arrow" onClick={() => sortData('city','ascending')}></i><i className="fa-solid fa-caret-down table__col-arrow" onClick={() => sortData('city','descending')}></i></div></div></th>
                    <th><div className='table__col'>State<div className='table__col-block'><i className="fa-solid fa-caret-up table__col-arrow" onClick={() => sortData('state','ascending')}></i><i className="fa-solid fa-caret-down table__col-arrow" onClick={() => sortData('state','descending')}></i></div></div></th>
                    <th><div className='table__col'>Zip Code<div className='table__col-block'><i className="fa-solid fa-caret-up table__col-arrow" onClick={() => sortData('zipCode','ascending')}></i><i className="fa-solid fa-caret-down table__col-arrow" onClick={() => sortData('zipCode','descending')}></i></div></div></th>
                </tr>
            </thead>
            <tbody>
                {sortedEmployees.slice(currentLine,(currentLine+numElementsRef.current.value)).map((employee, index) => (
                    <tr key={index}>
                        <td>{employee.firstname}</td>
                        <td>{employee.lastname}</td>
                        <td>{employee.dateBirthday}</td>
                        <td>{employee.dateStart}</td>
                        <td>{employee.departement}</td>
                        <td>{employee.street}</td>
                        <td>{employee.city}</td>
                        <td>{employee.state}</td>
                        <td>{employee.zipCode}</td>
                    </tr>
                ))}
            </tbody>
        </table>
        {renderPagination()}
        </>
    );
}
