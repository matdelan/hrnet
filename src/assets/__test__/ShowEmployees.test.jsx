
import { render, screen, fireEvent, waitFor, renderHook } from '@testing-library/react';
import { useRef } from 'react';
import { describe, it, expect, vi } from 'vitest';
import ShowEmployees from '../component/showemployees/ShowEmployees';
import { mockListEmployee } from '../__mocks__/listEmployee';

describe('<ShowEmployees>', () => {
    it('should render correctly', () => {
            const {container} = render( 
              <ShowEmployees employees={mockListEmployee.data} />
        )
        expect(container).toMatchSnapshot()
    })
})