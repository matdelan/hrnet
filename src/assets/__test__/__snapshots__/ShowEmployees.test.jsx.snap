// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ShowEmployees> > should render correctly 1`] = `
<div>
  <div
    class="table__header"
  >
    <div
      class="table__header-elem"
    >
      <label
        for="elementsPage"
      >
        Éléments par page : 
      </label>
      <select
        id="elementsPage"
      >
        <option
          value="10"
        >
          10
        </option>
        <option
          value="20"
        >
          20
        </option>
        <option
          value="50"
        >
          50
        </option>
        <option
          value="100"
        >
          100
        </option>
      </select>
    </div>
    <div
      class="table__header-elem"
    >
      <label
        for="search"
      >
        Search :
      </label>
      <input
        id="search"
        type="text"
      />
    </div>
  </div>
  <table>
    <thead>
      <tr
        class="table__col-gen"
      >
        <th>
          <div
            class="table__col"
          >
            First Name
            <span
              class="table__col-block"
            >
              <i
                class="fa-solid fa-caret-up table__col-arrow"
              />
              <i
                class="fa-solid fa-caret-down table__col-arrow"
              />
            </span>
          </div>
        </th>
        <th>
          <div
            class="table__col"
          >
            Last Name
            <div
              class="table__col-block"
            >
              <i
                class="fa-solid fa-caret-up table__col-arrow"
              />
              <i
                class="fa-solid fa-caret-down table__col-arrow"
              />
            </div>
          </div>
        </th>
        <th>
          <div
            class="table__col"
          >
            Date of Birth
            <div
              class="table__col-block"
            >
              <i
                class="fa-solid fa-caret-up table__col-arrow"
              />
              <i
                class="fa-solid fa-caret-down table__col-arrow"
              />
            </div>
          </div>
        </th>
        <th>
          <div
            class="table__col"
          >
            Start Date
            <div
              class="table__col-block"
            >
              <i
                class="fa-solid fa-caret-up table__col-arrow"
              />
              <i
                class="fa-solid fa-caret-down table__col-arrow"
              />
            </div>
          </div>
        </th>
        <th>
          <div
            class="table__col"
          >
            Department
            <div
              class="table__col-block"
            >
              <i
                class="fa-solid fa-caret-up table__col-arrow"
              />
              <i
                class="fa-solid fa-caret-down table__col-arrow"
              />
            </div>
          </div>
        </th>
        <th>
          <div
            class="table__col"
          >
            Street
            <div
              class="table__col-block"
            >
              <i
                class="fa-solid fa-caret-up table__col-arrow"
              />
              <i
                class="fa-solid fa-caret-down table__col-arrow"
              />
            </div>
          </div>
        </th>
        <th>
          <div
            class="table__col"
          >
            City
            <div
              class="table__col-block"
            >
               
              <i
                class="fa-solid fa-caret-up table__col-arrow"
              />
              <i
                class="fa-solid fa-caret-down table__col-arrow"
              />
            </div>
          </div>
        </th>
        <th>
          <div
            class="table__col"
          >
            State
            <div
              class="table__col-block"
            >
              <i
                class="fa-solid fa-caret-up table__col-arrow"
              />
              <i
                class="fa-solid fa-caret-down table__col-arrow"
              />
            </div>
          </div>
        </th>
        <th>
          <div
            class="table__col"
          >
            Zip Code
            <div
              class="table__col-block"
            >
              <i
                class="fa-solid fa-caret-up table__col-arrow"
              />
              <i
                class="fa-solid fa-caret-down table__col-arrow"
              />
            </div>
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          John
        </td>
        <td>
          Doe
        </td>
        <td>
          1990-01-01
        </td>
        <td>
          2022-09-01
        </td>
        <td />
        <td>
          123 Main St
        </td>
        <td>
          Springfield
        </td>
        <td>
          Illinois
        </td>
        <td>
          62704
        </td>
      </tr>
      <tr>
        <td>
          Johna
        </td>
        <td>
          aDoe
        </td>
        <td>
          1990-01-01
        </td>
        <td>
          2022-09-01
        </td>
        <td />
        <td>
          123 Main St
        </td>
        <td>
          Springfield
        </td>
        <td>
          Illinois
        </td>
        <td>
          62704
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="table__footer"
  >
    <div>
       Showing 
      1
       of 
      2
       of 
      2
        
    </div>
    <div
      class="table__pagination"
    >
      <div
        class="table__pagination-space"
      >
        Previous
      </div>
      <button
        class="activePage"
      >
        1
      </button>
      <div
        class="table__pagination-space"
      >
        Next
      </div>
    </div>
  </div>
</div>
`;
